providers/src/airflow/providers/edge/version_compat.py

```diff
@@ -0,0 +1,38 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+# NOTE! THIS FILE IS COPIED MANUALLY IN OTHER PROVIDERS DELIBERATELY TO AVOID ADDING UNNECESSARY
+# DEPENDENCIES BETWEEN PROVIDERS. IF YOU WANT TO ADD CONDITIONAL CODE IN YOUR PROVIDER THAT DEPENDS
+# ON AIRFLOW VERSION, PLEASE COPY THIS FILE TO THE ROOT PACKAGE OF YOUR PROVIDER AND IMPORT
+# THOSE CONSTANTS FROM IT RATHER THAN IMPORTING THEM FROM ANOTHER PROVIDER OR TEST CODE
+#
+
+from __future__ import annotations
+
+from packaging.version import Version
+
+
+def _get_base_airflow_version() -> Version:
+    from airflow import __version__
+
+    return Version(Version(__version__).base_version)
```

Comment:
We don't need base version do we? It will compare fine the way it is won't it?

```suggestion
def _get_base_airflow_version() -> Version:
    from airflow import __version__

    returnVersion(__version__)
```

Reply:
We do .. This is the main reason why we need to discuss all of that at all. The whole reason why we do this weird `Version(Version.base_version)` is because default comparision rules do not work properly for pre-release version and if we do not do it, checking for features available will break with pre-relase versions. 

I explained it earlier today in https://lists.apache.org/thread/h0s7ypdr05j10jw88b3svj57o4ml1brl - also there is this discussion in PyPI that explains why we have to do it https://github.com/pypa/pip/issues/13089  

Basically it's because of this paragraph in PEP-440:

>  Within a numeric release (1.0, 2.7.3), the following suffixes are
permitted and MUST be ordered as shown:
> .devN, aN, bN, rcN, <no suffix>, .postN

This paragraph in PEP 440 and it's implementation in `packaging` means that we have to do this weird looking double version retrieval, because in order to check "is feature available in this version" - we need to basically drop pre-release part of the version (which `Version(Version(__version__).base_version)` does.

Reply:
This is actually the **only** reason we have this discussion at all - anyone else could (like you) made the same mistake and check against `Version(_version_)`. And this is exactly why we have to do hand-holding / guide-railing - because it would break if you do it. For example if you want to do:

```python
from airflow __version__ 
if Version(__version__) > Version("3.0.0"):
  # do something only on Airflow 3
```

This condition is `False` on Airflow 3.0.0.dev0, 3.0.0.beta1, 3.0.0.rc1 ... this is the **gist** of the problem we are discussing.

Reply:
```python
>>> from packaging.version import Version
>>> Version("3.0.0rc1") >= Version("3.0.0")
False
>>> Version(Version("3.0.0rc1").base_version) >= Version("3.0.0")
True
>>>
```

Reply:
Ah k! I wonder if we should have a `from airflow import VERSION_INFO` (named to mirror `sys.version_info`) and then we could do `VERSION_TUPLE >= (3,0,0)`

Doesn't help us for any old checks though. And largely just stylistic.

Reply:
I'd say`_get_base_version` is a better abstraction for it, but yeah. Agree on both accounts - stilistic and does not help for old version checks. 

I can live with both. I just want to make sure for the next 8 months (and beyond) we do not have to deal with this :)

Reply:
Thought about it a bit - and I like the tuple approach more as it does not require top-level imports, which is always a good thing.  

Also this way I think it would satisfy @dstandish approach where he **only** would like to have `get_airflow_version` (or my equivalent `get_base_airflow_version_tuple` method) that could be easily used without using any predefined constants.

Say - you could inline this code then much easier (no additional Version import needed in the provider code):

```
get_airflow_base_version_tuple() > (3, 0, 0) 
```

if you do not want to use the constants (though I think constants are a bit better - mostly because it will be easier to remove them as we move the minimum version of airflow up. One benefit of that could be that every time we add new version of Airflow we woud not have to add a new constant (I think @dstandish  - this is what you really wanted if I read all your past comment and try to extract the real concerns you had).

I propose to merge this one (with the test version_constant  separation in #44770  as prerequisite), and maybe as next step if others  agree - we could change all provider's code to use such inlined version and change instruction on how we should do it). I am happy to make a draft PR for this approach @dstandish - to see how it would look up. @ashb  @gopidesupavan @ashb @jscheffl and others - let me know what you think and whether it is worth to do another follow-up - but let's close this one first to not add confusion.

From the very beginning I was not into "let's do it this way" but "let's address the issue we have at hand - for now and the future" - so I am happy to implement it in the way we can all feel comfortable with.  

I know it took a bit of many people's time, but I think it's important to solve it for the next 8 months to come - while I was making this one pass, I just found - out the 7th (!) way that such check was done inside the openlineage (with even differently named constant used to hold airflow version).

Reply:
I do not have a strong opinion atm and prefer a simple and consolidated approach. Good to merge in my view and also if there is a better solution OK to change it later. Thanks for the energy making it better!

Reply:
Yeah. The way it is changed now, makes it *easy* to change it as a follow-up - no more "common" code impact, dependencies etc., just common convention applied everywhere - and it can be easily changed now as we are guarding it a bit better and telling people what to do to add more compatibility code, so there is a chance both contributors and maintainers will stick to the convention. 

Also - if the #44607 from @dstandish will follow the same (or introduce new - consistently) convention, the pre-commit code we have will guard that the new code added to airflow will not be used until we decide it's OK to use it (basically min-airlfow-version == 2.11) in this case.

Reply:
BTW. I've learned from my friend a very obscene saying in Polish - freely translated `even if sh**tty, it's better  to be consistent` - in Polish it even rhmes ("Lepiej ch**owo ale jednakowo") - which I think applies here very well.

I basically don't really care how we do it, as long as we do it consistently and set the stage to keep it consistent (rather than in 7 different ways).

Reply:
I am okay with this changes and to merge, I am also with jens comment , at the moment it is simple and solves our problems.
And i really appreciate your effort :)

Reply:
> And i really appreciate your effort :)

That's my `honey-badger` genes in full swing (cc: @dstandish )